packages/grpc-js/src/load-balancer-eds.ts

```diff
@@ -0,0 +1,313 @@
+      if (highestOldPriority === Infinity) {
+        /* No existing priority at or below the same number as the priority we
+         * are looking at had any of the localities in this priority. So, we
+         * use a new name. */
+        newPriorityName = `child${this.nextPriorityChildNumber++}`;
+      } else {
+        const newName = this.priorityNames[highestOldPriority];
+        if (newPriorityNames.indexOf(newName) < 0) {
+          newPriorityName = newName;
+        } else {
+          newPriorityName = `child${this.nextPriorityChildNumber++}`;
+        }
+      }
+      newPriorityNames[priority] = newPriorityName;
+
+      const childTargets: Map<string, WeightedTarget> = new Map<
+        string,
+        WeightedTarget
+      >();
+      for (const localityObj of localityArray) {
+        childTargets.set(localityToName(localityObj.locality), {
+          weight: localityObj.weight,
+          /* TODO(murgatroid99): Insert an lrs config around the round_robin
+           * config after implementing lrs */
+          /* Use the endpoint picking policy from the config, default to
+           * round_robin. */
+          child_policy: [
+            ...this.lastestConfig.eds.endpointPickingPolicy,
+            { name: 'round_robin', round_robin: {} },
+          ],
+        });
+        for (const address of localityObj.addresses) {
+          addressList.push({
+            localityPath: [
+              newPriorityName,
+              localityToName(localityObj.locality),
+            ],
+            ...address,
+          });
+        }
+      }
+
+      priorityChildren.set(newPriorityName, {
+        config: [
+          {
+            name: 'weighted_target',
+            weighted_target: {
+              targets: childTargets,
+            },
+          },
+        ],
+      });
+    }
+    const childConfig: PriorityLoadBalancingConfig = {
+      name: 'priority',
+      priority: {
+        children: priorityChildren,
+        /* Contract the priority names array if it is sparse. This config only
+         * cares about the order of priorities, not their specific numbers */
+        priorities: newPriorityNames.filter((value) => value !== undefined),
+      },
+    };
+    this.childBalancer.updateAddressList(
+      addressList,
+      childConfig,
+      this.latestAttributes
+    );
+
+    this.localityPriorities = newLocalityPriorities;
+    this.priorityNames = newPriorityNames;
+  }
+
+  updateAddressList(
+    addressList: SubchannelAddress[],
+    lbConfig: LoadBalancingConfig,
+    attributes: { [key: string]: unknown }
+  ): void {
+    if (!isEdsLoadBalancingConfig(lbConfig)) {
+      return;
+    }
+    if (!(attributes.xdsClient instanceof XdsClient)) {
+      return;
+    }
+    this.lastestConfig = lbConfig;
+    this.latestAttributes = attributes;
+    this.xdsClient = attributes.xdsClient;
+    this.edsServiceName = lbConfig.eds.edsServiceName ?? lbConfig.eds.cluster;
+
+    if (!this.isWatcherActive) {
```

Comment:
As discussed offline, what if this function `updateAddressList` is called with a different `lbConfig.eds.edsServiceName`? After the first call, `this.isWatcherActive` is set to true. Will the 2nd call not get `this.xdsClient.addEndpointWatcher()` called?

Or maybe I have misunderstood the semantics - maybe `edsServiceName` won't change?

Reply:
I have changed this behavior to handle the name changing.